server/src/test/java/org/elasticsearch/repositories/blobstore/BlobStoreRepositoryTests.java

```diff
@@ -798,4 +752,202 @@ public void testUuidCreationLogging() {
+                final var expectedBlobsToDelete = new HashSet<String>();
+
+                final var countDownLatch = new CountDownLatch(1);
+                try (var refs = new RefCountingRunnable(countDownLatch::countDown)) {
+                    for (int index = between(200, 1000); index > 0; index--) {
+                        final var indexId = new IndexId(randomIdentifier(), randomUUID());
+                        for (int shard = between(5, 30); shard > 0; shard--) {
+                            final var shardId = shard;
+                            final var shardGeneration = new ShardGeneration(randomUUID());
+                            expectedShardGenerations.put(indexId, shard, shardGeneration);
+                            final var blobsToDelete = generateRandomBlobsToDelete(20, 100);
+
+                            final var indexPath = repo.basePath()
+                                .add("indices")
+                                .add(indexId.getId())
+                                .add(Integer.toString(shard))
+                                .buildAsString();
+                            for (final var blobToDelete : blobsToDelete) {
+                                expectedBlobsToDelete.add(indexPath + blobToDelete);
+                            }
+
+                            repo.threadPool().generic().execute(ActionRunnable.run(refs.acquireListener(), () -> {
+                                shardBlobsToDelete.addShardDeleteResult(indexId, shardId, shardGeneration, blobsToDelete);
+                            }));
+                        }
+                    }
+                }
+
+                safeAwait(countDownLatch);
+                assertEquals(expectedShardGenerations.build(), shardBlobsToDelete.getUpdatedShardGenerations());
+                shardBlobsToDelete.getBlobPaths().forEachRemaining(s -> assertTrue(expectedBlobsToDelete.remove(s)));
+                // Some writes were truncated, and so not all blobPaths are returned
+                assertFalse(expectedBlobsToDelete.isEmpty());
+                mockLog.assertAllExpectationsMatched();
+            }
+        }
+    }
+
+    /*
+        There's no heap space, so we don't even attempt to write
+     */
+    @TestLogging(reason = "test includes assertions about logging", value = "org.elasticsearch.repositories.blobstore:WARN")
+    public void testShardBlobsToDeleteWithOutHeapSpace() {
+        IndexId indexId = new IndexId(randomIdentifier(), randomUUID());
+        int shardId = between(1, 30);
+        Collection<String> blobIds = generateRandomBlobsToDelete(0, 20);
+        ShardGeneration shardGeneration = new ShardGeneration(randomUUID());
+
+        final var expectedShardGenerations = ShardGenerations.builder().put(indexId, shardId, shardGeneration).build();
+
+        Settings.Builder settings = Settings.builder().put("repositories.blobstore.max_shard_delete_results_size", "0b");
```

Comment:
Likewise here I think this can just be folded into `testShardBlobsToDeleteWithLimitedHeapSpace`.